### PR TITLE
Disconnect existing gateway connections

### LINKED ISSUE #1987: New connections overwrite existing connections
#### Summary

New connections overwrite existing connections

#### Steps to Reproduce

1. Connect gateway over UDP
2. Connect another gateway with the same EUI
3. Disconnect any of the gateways
4. The gateway by UID gets removed from a map which is used for downlink, so downlink is broken

#### What do you see now?

Gateway Server maps the connections by UID and removes entries on gateway disconnects, while there may be another gateway connected. This can be due to an unclean disconnect or anything else. References https://github.com/TheThingsNetwork/lorawan-stack/issues/1729 and probably #1730 (and probably https://github.com/TheThingsIndustries/lorawan-stack/issues/1735). 

#### What do you want to see instead?

Actively disconnect existing connections when a gateway connects.

#### How do you propose to implement this?

https://github.com/TheThingsNetwork/lorawan-stack/blob/master/pkg/gatewayserver/gatewayserver.go#L415

Here we probably want a loop that does a `LoadOrStore()`, and as long as there's a connection loaded, disconnect it and wait for that connection. You can close a signaling channel here: https://github.com/TheThingsNetwork/lorawan-stack/blob/master/pkg/gatewayserver/gatewayserver.go#L476

That waiting is necessary to clean up all upstream handling of the old connection and ensure that the `Delete()` from the map is the concerning connection, and not already the new connection, still breaking downlink.

#### Can you do this yourself and submit a Pull Request?

Can review, can do

### PR DESCRIPTION
**Pull request by johanstokking**
#### Summary

Closes #1987

#### Changes

- Disconnect existing gateway connections before handling a new connection. This cleans up all existing Basic Station, MQTT and gRPC connections and is tested. UDP is not necessary since it maps by EUI and has its own connection expiry mechanism

#### Notes for Reviewers

n/a

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [x] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by johanstokking** [comment]
References #1729 #1730 https://github.com/TheThingsIndustries/lorawan-stack/issues/1735

File: pkg/gatewayserver/gatewayserver.go [hunk 1]
```diff
@@ -412,10 +419,23 @@ func (gs *GatewayServer) Connect(ctx context.Context, frontend io.Frontend, ids
 	if err != nil {
 		return nil, err
 	}
-	gs.connections.Store(uid, conn)
+	connEntry := connectionEntry{
+		Connection:   conn,
+		upstreamDone: make(chan struct{}),
+	}
+	for existing, exists := gs.connections.LoadOrStore(uid, connEntry); exists; existing, exists = gs.connections.LoadOrStore(uid, connEntry) {
```
Comment (KrishnaIyer) [inline pkg/gatewayserver/gatewayserver.go]: I see a potential deadlock here. 
- There's Conn A.
- Conn B now replaces Conn A and blocks until A is closed.
- Now Conn C will replace Conn B and force it to close.
- But if Conn B continues the loop before Conn C closes it  via `L431`, then Conn B again replaces Conn C and try to close it.

It's kinda rare but theoretically possible. How about disallowing a second (or multiple) connection and also closing the existing connection?
Then when the gateway retries we can have a clean connect from the beginning?

Comment (johanstokking) [reply pkg/gatewayserver/gatewayserver.go]: >But if Conn B continues the loop before Conn C closes it 

This doesn't happen. `LoadOrStore()` is atomic; if it gets loaded, that existing connection gets closed and we can wait for that. If it doesn't get loaded, the connection is by definition stored, the loop breaks and the upstream handler gets started. Even if the upstream handler is not up yet, connection C can safely wait for the channel close, as connection B will definitely start that goroutine.

**Comment by johanstokking** [comment]
@neoaggelos indeed the front-ends stop receiving traffic when the connection gets closed.

As far as there is buffered traffic: closing the connection context stops the upstream handlers as soon as their current message in flight is done. Buffered messages from the old connection are discarded and they’ll get garbage collected because everything goes out of scope.

In fact, the for-loop only continues when all upstream handlers are done so nothing happens in parallel. It may be that the new connection needs to wait for one uplink message in flight to be published to Packet Broker, for example, but that’s the worst that can happen, and also that context gets cancelled.

Or do you mean something else?

**Comment by johanstokking** [comment]
Ah I see what you mean. Yes, a read on a socket returns an error when the connection gets closed.

**Pull request merged**